lib/rules/function-calc-no-unspaced-operator/index.mjs

```diff
@@ -65,197 +86,486 @@ const rule = (primary, _secondaryOptions, context) => {
+const OPERAND_TOKEN_TYPES = new Set([
+	TokenType.Number,
+	TokenType.Dimension,
+	TokenType.Percentage,
+	TokenType.Ident,
+]);
+
+/**
+ * @param {ComponentValue | undefined} node
+ * @returns {boolean}
+ */
+function isOperandNode(node) {
+	if (isSimpleBlockNode(node)) return true;
+
+	if (isFunctionNode(node)) {
+		const name = node.getName().toLowerCase();
+
+		if (mathFunctions.has(name) || name === 'var') return true;
+
+		return false;
+	}
+
+	if (!isTokenNode(node)) return false;
+
+	return OPERAND_TOKEN_TYPES.has(node.value[0]);
+}
+
+/**
+ * NOTE: this is messy.
+ *
+ * @typedef {{
+ *   firstOperand: ComponentValue,
+ *   before: Array<ComponentValue>,
+ *   secondOperand: ComponentValue,
+ *   after: Array<ComponentValue>,
+ *   operator: ComponentValue | undefined,
+ *   operatorToken: import ('@csstools/css-tokenizer').TokenDelim | undefined,
+ * }} Operation
+ *
+ * @typedef {{
+ *   firstOperand: ComponentValue,
+ *   before: Array<ComponentValue>,
+ *   secondOperand: ComponentValue,
+ *   after: Array<ComponentValue>,
+ *   operator: ComponentValue,
+ *   operatorToken: import ('@csstools/css-tokenizer').TokenDelim,
+ * }} CompleteOperation
+ */
+
 /**
- * @param {string} str
- * @param {number} index
- * @param {string} char
+ * @param {Operation} operation
+ * @returns {operation is CompleteOperation}
  */
-function insertCharAtIndex(str, index, char) {
-	return str.slice(0, index) + char + str.slice(index, str.length);
+function isCompleteOperation(operation) {
+	return (
+		Boolean(operation.firstOperand) &&
+		Boolean(operation.secondOperand) &&
+		Boolean(operation.operator) &&
+		Boolean(operation.operatorToken)
+	);
 }
 
 /**
- * @param {import('postcss-value-parser').Node | undefined} node
- * @returns {node is import('postcss-value-parser').SpaceNode}
+ * @param {Operation} operation
+ * @returns {operation is Operation}
  */
-function isSingleSpace(node) {
-	return node != null && isSpace(node) && node.value === ' ';
+function isOperationWithoutOperator(operation) {
+	return (
+		Boolean(operation.firstOperand) &&
+		Boolean(operation.secondOperand) &&
+		!operation.operator &&
+		!operation.operatorToken
+	);
 }
 
 /**
- * @param {import('postcss-value-parser').Node | undefined} node
- * @param {Set<string>} [operators]
- * @returns {node is import('postcss-value-parser').WordNode}
+ * @param {ContainerNode} container
+ * @param {number} cursor
+ * @returns {[Operation | undefined, number]}
  */
-function isOperator(node, operators = OPERATORS) {
-	return node != null && isWord(node) && operators.has(node.value);
+function parseOperation(container, cursor) {
+	/** @type {ComponentValue | undefined} */
+	let firstOperand = undefined;
+	/** @type {ComponentValue | undefined} */
+	let secondOperand = undefined;
+	/** @type {Array<ComponentValue>} */
+	const before = [];
+	/** @type {Array<ComponentValue>} */
+	const after = [];
+	/** @type {ComponentValue | undefined} */
+	let operator = undefined;
```

Comment:
style
refactor

```suggestion
	/** @type {Array.<ComponentValue | undefined>} */
	let [firstOperand, secondOperand, operator] = [];
	/** @type {Array<ComponentValue>} */
	const before = [];
	/** @type {Array<ComponentValue>} */
	const after = [];
```

Reply:
To be honest, I'm against this idea because `undefined` in the `ComponentValue | undefined` type is hidden and unclear.